Formula/chapel.rb

```diff
@@ -15,19 +16,39 @@ class Chapel < Formula
   end
 
   depends_on "gmp"
-  depends_on "python@3.10"
+  # `chapel` has scripts that use python on PATH (e.g. via `command -v python3`).
+  # So, it needs to depend on the currently linked Homebrew Python version.
+  # TODO: remove from versioned_dependencies_conflicts_allowlist when python
+  # dependency version matches that of all llvm dependencies.
+  depends_on "python@3.9"
```

Comment:
I guess another option would be to overwrite the script(s), though I don't know how many scripts need to be changed and if any shebangs need to be updated (there are some occurrences of `#!/usr/bin/env python3`).

Main script that is impacting Linux is probably https://github.com/chapel-lang/chapel/blob/1.26.0/util/config/find-python.sh which checks for `python3` `python` `python2` on PATH. Not sure if it will break anything if we just overwrite it with 
```rb
File.open(libexec/"util/config/find-python.sh", "w") do |f|
  f.write <<~EOS
    #!/bin/sh
    echo "#{which("python3")}"
  EOS
end
```

Currently safer to just use linked Homebrew python, but this is another option to consider.

Reply:
Is there a reason we can't just use `ENV.prepend_path "PATH" Formula["python@3.10"].opt_bin` so that `python3` is found there first?  Or does that not work?

As we were discussing on Slack, it sounds like we're leaning heavily in favor of just temporarily adding formulae to the `versioned_dependencies_conflicts_allowlist` for Python version conflicts rather than resorting to these more complex fixes.

Reply:
It depends whether installed `chapel` scripts are usable by users. If we just modify `PATH` in test, that means users have to do the same to use the scripts. In this case, we may as just make Python a build/test dependency as the parts that need python won't work with `python@3.10` dependency anyway without user intervention.

If the scripts are actually useful, this then means we need to add a caveat that users need to install & update `PATH` (as this is a Homebrew installation behavior), which I feel is a bit awkward choice. Here, I would just recommend using `python@3.9` instead.

Reply:
From basic help test, looks like Chapel isn't even usable if `python3` is not available on PATH:
```console
$ chpl --help
python3 not found - please install python3 and add it to PATH
/home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/util/printchplenv: 11: /home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/util/printchplenv: python3: not found
error: 'CHPL_HOME=/home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec CHPL_RUNTIME_INCL=/home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/runtime/include CHPL_RUNTIME_LIB=/home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/lib CHPL_THIRD_PARTY=/home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/third-party CHPLENV_SKIP_HOST=true CHPLENV_SUPPRESS_WARNINGS=true /home/linuxbrew/.linuxbrew/Cellar/chapel/1.26.0/libexec/util/printchplenv --all --internal --no-tidy --simple' did not run successfully
```

For bin symlinked scripts like `chpl`, we could do a `PATH` override. 

The question is whether any `libexec` scripts are useful. From general Chapel documentation, I see them mention running `$CHPL_HOME/util/...` (`CHPL_HOME=libexec`) scripts that are not going to run on Linux Homebrew install and currently run using `python3.8` on macOS 10.15+ (`python2` on macOS <= 10.14).

Reply:
This does sound more complicated than it first appears.  This seems to be rare case where this cannot reasonably be migrated to `python@3.10` until that becomes the linked version instead of `python@3.9`.  

The only remaining issue is switching LLVM versions.  If the only reason we need to switch this to `llvm@12` is to avoid conflicting versioned dependencies for `python@3.9`/`python@3.10`, we should just add it to the allowlist.  I'm assuming the reason we changed it to `llvm@13` in the LLVM PR was because it's not compatible with LLVM 14, but I don't think we should have to switch to `llvm@12` unless it actually doesn't work with LLVM 13.

Reply:
Based on this I think it makes sense then to just do the `python@3.9` downgrade here and add it to the versioned dependency allowlist and drop the `llvm@12` downgrade.  In the LLVM PR we can then just switch it `llvm@13`, with a comment indicating that can be upgraded to `llvm` at the next release.  

It turns out we've never been checking for versioned dependency conflicts on Linux and nothing crazy has happened.  The only scenario I can see being a problem with conflicting Python versions is if somehow a formulae ended up dynamically linking a native binary to both `libpython3.9` and `libpython3.10` and this resulted in some collisions in symbol names.  This seems exceptionally unlikely to occur so I think we should just be more permissive about conflicting versioned Python dependencies so as not to block migrations.

Reply:
That's fine with me. I think if we extract LLDB out of the `llvm` formula, we could reduce these Python conflict scenarios, but we can consider that after we finish LLVM 14 PR.

I think the more likely scenario is if the build environment picks up a particular formula but RPATH causes us to pick an alternative. This is what caused the `pdfsandwich` failure where `libomp` (v14) & `llvm` (v13) were both in dependency tree (though this particular example wouldn't have been caught even if we audited since we don't detect non-versioned formulae with alternative/drop-in libraries).

I guess another scenario is if build system detects the "wrong" python version, but should be unlikely assuming we prioritize the direct dependents over indirect dependents on `PATH` and other environment variables. Haven't checked `brew` logic though.

Reply:
I definitely agree that we can't ignore versioned dependency conflicts when there is a risk of mismatched linkage like that scenario you described with `pdfsandwich/imagemagick`.  But in the case of Python, very few formulae are going to link to `libpython` to start with so the chance of linking two versions of it seems low.

We definitely need to explore the options for splitting LLDB and `libc++` from LLVM at some point as LLDB probably has no dependents, and nothing is supposed to link to `libc++` that comes with LLVM - C++ formulae should either `libstdc++` on Linux or the host `libc++` on macOS.

I haven't seen a scenario yet where the wrong Python gets picked when we explicitly depend on a particular version.  I'm pretty sure we do prioritize direct dependencies in the `PATH` over indirect ones but as you said it's probably worth checking that.